modules/caddyhttp/caddyhttp.go

```diff
@@ -245,10 +245,32 @@ func SanitizedPathJoin(root, reqPath string) string {
 	return path
 }
 
-// CleanPath cleans path p according to path.Clean() except that
-// it preserves trailing slashes, since those are significant
-// in our application.
-func CleanPath(p string) string {
+// CleanPath cleans path p according to path.Clean(), but only
+// merges repeated slashes if collapseSlashes is true, and always
+// preserves trailing slashes.
+func CleanPath(p string, collapseSlashes bool) string {
+	if collapseSlashes {
+		return cleanPath(p)
+	}
+
+	// replace repeated slashes with an invalid/impossible URI character
+	// to avoid collisions; then clean the path, then replace them back
+	const tmpSlash = 0xff
+	var sb strings.Builder
+	for i, ch := range p {
+		if ch == '/' && i > 0 && (p[i-1] == '/' || p[i-1] == tmpSlash) {
```

Comment:
I guess `p` should not contain any `tmpSlash`?

Reply:
Ah yeah, it's not intuitive. It does contain `tmpSlash` because we put it in there:

```
1. /foo///bar
2. /foo/0xFF/bar
3. /foo/0xFF0xFFbar
```

On that third iteration, the 1-lookbehind sees 0xFF because we barely replaced the `/` with `0xFF`.

Reply:
I think `p` is immutable, while `sb` grows:

| iteration | `p` | `sb` |
| --- | --- | --- |
| 1 | /foo///bar | /foo/ |
| 2 | /foo///bar | /foo/0xFF |
| 3 | /foo///bar | /foo/0xFF0xFF |

Reply:
Oh... that's right. (It was late when I wrote that and it's late now.) Why does this work then?